DataDog/chaos-controller
# Dynamic Targeting

## What does this PR do?

- [X] Adds new functionality
- [ ] Alters existing functionality
- [ ] Improves documentation or testing

### Current state of targeting
- As of right now, the controller does not continue targeting in time; at appliance, the disruption selects its targets once and will stick to them. 
- This is a very useful behaviour for a couple of reasons:
  - it makes the monitoring of targeting simpler to manage, more stable and transparent
  - it limits the impact of a disruption as new pods (which could compensate for the disruption) will not be targeted

Non-dynamic targeting will stay the default behaviour of the controller for safety and simplicity reason, as it fits most use cases. However, it is also a limit for some situations: a notorious case being the constantly-deploying services. If a service's pods are rolled out every 5-10 minutes, it is impossible to target a service properly for an extended period of time, as new pods will be left intact. Hence the implementation of Dynamic Targeting within the `chaos-controller`

### New Feature: Dynamic Targeting
Dynamic Targeting will allow the controller to follow-up on a cluster/deployment/target/etc. state in order to update the disruption and maintain it's targeting objective. With dynamic targeting activated, a disruption will watch the cluster's activity in order to disrupt new targets (if required) and stop counting dead targets as targets (eg. a 100%-count disruption will constantly disrupt 100% of old and new fitting targets within the namespace)

To achieve this result, the controller watches over the disruption's given namespace and target label selector.

![Activity diagram for Dynamic Targeting (2)](https://user-images.githubusercontent.com/17198797/157055173-f4ab9d94-5c4d-419d-a08f-160fd41c5f23.png)
![Targets Update_Reconcile Logic (2)](https://user-images.githubusercontent.com/17198797/162756817-21b3f8f3-15e4-4e02-badc-c823f3f583ac.png)



## Code Quality Checklist

- [ ] My code is sufficiently commented and passes continuous integration checks.
- [X] I have signed my commit (see [Contributing Docs](../CONTRIBUTING.md)).

## Testing

- [ ] I leveraged continuous integration testing
    - [ ] by depending on existing `unit` tests or `end-to-end` tests.
    - [ ] by adding new `unit` tests or `end-to-end` tests.
- [X] I manually tested the following steps:
    - On minikube
        - Install the controller on minikube
        - Edit any pod disruption with `dynamicTargeting: true` and `count: 100%` (or `50%`)
        - `examples/demo.yaml`: Edit the `demo-curl` container CPU limit to 0.1 (currently 1)
        - Deploy the demo
        - apply the chosen disruption
        - edit the demo-curl amount of replicas to *X* (arbitrary number) -- apply
        - Observe the number of chaos-pods go to *X*
        - edit the demo-curl amount of replicas to *X/2* (lower number) -- apply
        - Observe the number of chaos-pods go to *X/2*
        - Start again with `dynamicTargeting: false` and this time the amount of pods will not change
    - [X] as a canary deployment to a cluster.
